File: js/data_structures/loop_link_list.js
// 循环链表
const LinkList = require('./link_list')

class LoopLinkList extends LinkList {
    constructor() {
        super();
        // 放置一个记录当前指针位置的索引
        this.curNode = this.head
        this.head.next = this.head
    }
    // 从当前索引向后移动n个位置
    move(n) {
        for (let i = 0; i < n; i++) {
            this.curNode = this.curNode.next
            // 跳过头节点
            if(this.curNode.item == 'head' && this.curNode.next.item != 'head') {
                this.curNode = this.curNode.next
            }
        }
    }
    // 回到头部
    moveToHead() {
        this.curNode = this.head
    }
    // 展示当前指针位置
    show() {
        return this.curNode.item
    }
    toArr() {
        let curNode = this.head
        const arr = [];

        while (curNode.next.item != 'head') {
            arr.push(curNode.next.item)
            curNode = curNode.next
        }

        return arr
    }
}

module.exports = LoopLinkList

File: py/data_structures/loop_linklist.py
# 循环链表
from data_structures.linklist import LinkList


# 循环链表实现
class LoopLinkList(LinkList):
    def __init__(self):
        super().__init__()
        # 构建循环链表
        # 链表指针
        self.cur_node = self.head
        self.head.next = self.head

    # 修改to_arr方法
    def to_arr(self):
        cur_node = self.head
        arr = []
        while cur_node.next.item != 'head':
            arr.append(cur_node.next.item)
            cur_node = cur_node.next
        return arr
    
    # 指针向前移动n个位置
    def move(self, n):
        i = 0
        while i < n:
            i += 1
            self.cur_node = self.cur_node.next
            # 处理移动到头部后，自动走到下一位
            if self.cur_node.item == 'head' and self.cur_node.next.item != 'head':
                self.cur_node = self.cur_node.next

    # 回到头部
    def move_to_head(self):
        self.cur_node = self.head

    def show(self):
        return self.cur_node
